lib/kube/proxy/transport_retryable.go

```diff
@@ -0,0 +1,272 @@
+		)
+		return func() {}
+	}
+
+	// Skip zero-length bodies.
+	if req.ContentLength == 0 {
+		return func() {}
+	}
+
+	// Request body is too large to retry.
+	// Allows balanced use of memory budget across requests.
+	if req.ContentLength > rt.maxBufferPerRequest {
+		rt.log.InfoContext(req.Context(),
+			"Request body too large for retry buffer, request will not be retryable on GOAWAY",
+			"size", req.ContentLength,
+			"limit", rt.maxBufferPerRequest,
+			"method", req.Method,
+			"path", req.URL.Path,
+		)
+		return func() {}
+	}
+
+	// Acquire memory, blocks if unavailable.
+	if err := rt.semaphore.Acquire(req.Context(), req.ContentLength); err != nil {
+		rt.log.InfoContext(req.Context(),
+			"Unable to acquire retry buffer semaphore, request will not be retryable on GOAWAY",
+			"size", req.ContentLength,
+			"method", req.Method,
+			"path", req.URL.Path,
+			"error", err,
+		)
+		return func() {}
+	}
+
+	rb := newRetryBuffer(req.Body, int(req.ContentLength))
+	req.Body = rb
+	req.GetBody = func() (io.ReadCloser, error) {
+		return rb.getBody()
+	}
+
+	releaseSemaphore := func() {
+		if err := req.Body.Close(); err != nil {
+			rt.log.DebugContext(req.Context(),
+				"Unable to close request body",
+				"size", req.ContentLength,
+				"method", req.Method,
+				"path", req.URL.Path,
+				"error", err,
+			)
+		}
+		rt.semaphore.Release(req.ContentLength)
+	}
+	return releaseSemaphore
+}
+
+// isUpgrade detects HTTP/1.1 protocol upgrades (WebSocket, SPDY) that won't
+// receive HTTP/2 GOAWAY frames.
+//
+// We cannot check req.Proto to detect HTTP/2 because modifyRequest() sets it
+// to "HTTP/1.1" for all outgoing requests. Instead, we rely on the fact that
+// all transports wrapped by retryableTransport are created via newH2Transport(),
+// which configures HTTP/2. Requests without Upgrade headers will use HTTP/2 and
+// may receive GOAWAY.
+func (rt *retryableTransport) isUpgrade(req *http.Request) bool {
+	return req.Header.Get("Connection") == "Upgrade" ||
+		req.Header.Get("Upgrade") != "" ||
+		req.Header.Get("X-Stream-Protocol-Version") != ""
+}
+
+// CloseIdleConnections closes idle connections in the inner transport.
+func (rt *retryableTransport) CloseIdleConnections() {
+	if ci, ok := rt.inner.(interface{ CloseIdleConnections() }); ok {
+		ci.CloseIdleConnections()
+	}
+}
+
+// retryBuffer incrementally reads from a source, incrementally writes
+// to a buffer, and gracefully completes reading during any early call to Close.
+type retryBuffer struct {
+	src     io.ReadCloser
+	buf     *bytes.Buffer
+	mu      sync.Mutex
+	cond    *sync.Cond
+	readErr error
+}
+
+func newRetryBuffer(source io.ReadCloser, capacity int) *retryBuffer {
+	type limitedReadCloser struct {
+		*io.LimitedReader
+		io.Closer
+	}
+	rb := &retryBuffer{
+		// Limit reading in case actual body size is
+		// greater than presented ContentLength.
+		src: &limitedReadCloser{
+			LimitedReader: io.LimitReader(source, int64(capacity)).(*io.LimitedReader),
```

Comment:
It looks like the fields of the LimitReader are public. Can we instantiate the struct directly instead of casting?
```suggestion
			LimitedReader: &io.LimitReader{R: source, N: int64(capacity)},
```